src/state.rs

```diff
@@ -0,0 +1,295 @@
+            self.current_package = package_name;
+            Ok(())
+        } else {
+            bail!("Package {package_name} not found")
+        }
+    }
+
+    #[inline]
+    pub fn get_current_package_name(&self) -> &SymbolRef {
+        &self.current_package
+    }
+
+    fn get_current_package(&self) -> &Package {
+        self.symbol_packages
+            .get(&self.current_package)
+            .expect("current package must be in the hashmap")
+    }
+
+    fn get_current_package_mut(&mut self) -> &mut Package {
+        self.symbol_packages
+            .get_mut(&self.current_package)
+            .expect("current package must be in the hashmap")
+    }
+
+    pub fn resolve(&self, symbol_name: &str) -> Option<&SymbolRef> {
+        self.get_current_package().resolve(symbol_name)
+    }
+
+    pub fn intern<A: AsRef<str>>(&mut self, symbol_name: A) -> SymbolRef {
+        self.get_current_package_mut()
+            .intern(String::from(symbol_name.as_ref()))
+    }
+
+    pub fn import(&mut self, symbols: &[SymbolRef]) -> Result<()> {
+        self.get_current_package_mut().import(symbols)
+    }
+
+    pub fn use_package(&mut self, package: &Package) -> Result<()> {
+        self.get_current_package_mut().use_package(package)
+    }
+
+    pub fn fmt_to_string(&self, symbol: &SymbolRef) -> String {
+        self.get_current_package().fmt_to_string(symbol)
+    }
+
+    fn intern_fold<A: AsRef<str>>(
+        &mut self,
+        init: SymbolRef,
+        path: &[A],
+        create_unknown_packges: bool,
+    ) -> Result<SymbolRef> {
+        path.iter()
+            .try_fold(init, |acc, s| match self.symbol_packages.get_mut(&acc) {
+                Some(package) => Ok(package.intern(String::from(s.as_ref()))),
+                None => {
+                    if create_unknown_packges {
+                        let mut package = Package::new(acc);
+                        let symbol = package.intern(String::from(s.as_ref()));
+                        self.add_package(package);
+                        Ok(symbol)
+                    } else {
+                        bail!("Package {acc} not found")
+                    }
+                }
+            })
+    }
+
+    #[inline]
+    pub fn intern_path<A: AsRef<str>>(
+        &mut self,
+        path: &[A],
+        keyword: bool,
+        create_unknown_packges: bool,
+    ) -> Result<SymbolRef> {
+        self.intern_fold(Symbol::root(keyword).into(), path, create_unknown_packges)
+    }
+
+    #[inline]
+    pub fn intern_relative_path<A: AsRef<str>>(
+        &mut self,
+        path: &[A],
+        create_unknown_packges: bool,
+    ) -> Result<SymbolRef> {
+        self.intern_fold(self.current_package.clone(), path, create_unknown_packges)
+    }
+
+    #[inline]
+    pub fn minimal() -> Self {
+        Self::new_with_package(Package::new(SymbolRef::new(Symbol::root_sym())))
+    }
+
+    pub fn init_lurk_state() -> Self {
```

Comment:
Init feels like an imperative, so maybe this could be called `initial_lurk_state()` — since you're not 'initializing' something.

Reply:
There's `init` and `initial` as separate functions. `init` does some work and initializes the state with the right structure. `initial` is the stable reference to an initialized state

Reply:
Got it, thanks — was reviewing top-to-bottom.